flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java

```diff
@@ -411,11 +411,22 @@ private void registerMetricSync() {
      * Logic needs to be invoked by write AND flush since we support various semantics.
      */
     private void checkAsyncException() throws IOException {
-        // reset this exception since we could close the writer later on
         Exception e = asyncProducerException;
         if (e != null) {
-
-            asyncProducerException = null;
+            // In old version, asyncProducerException will be set to null here, which causes another
+            // bug [FLINK-35749]
+            // once asyncProducerException is set to null here, the MailboxExecutor thread in
+            // WriterCallback.onCompletion may also invoke this method checkAsyncException and get
+            // chance to make
+            // asyncProducerException be null before the function flush() and write() seeing the
+            // exception.
+            // After that, the flush and writer functions can continue to write next message batch
+            // to kafka, the old
+            // in-flight message batch with exceptions get lost.
+            // We hope that once some exceptions thrown during sending kafka, KafkaWriter could be
+            // fast-failed and
+            // trigger global restarting, then asyncProducerException will be initialized as null
+            // again.
```

Comment:
I don't think this change is necessary at all. You can see that we now have inflated error count metrics in all the tests that you adjusted (and should be reverted as well).

The issue is not that we reset the exception. It's that it doesn't bubble up (see below). We actually have a clear contract on the volatile `asyncProducerException`: it's set in the `onCompletion` and read+reset in the main thread (which is the mailbox thread). So we never have a chance to concurrently reset the exception and lose it somehow.

Reply:
Thanks for explaining, and I want to discuss this part with you for more detail.

As we can see, the read-reset-throw logic is in checkAsyncException(), and there 4 parent method may invoke checkAsyncException():
1. writer()  2. flush()  3. close()  4. WriterCallback#onCompletion

I agree with "we never have a chance to concurrently reset the exception ", but 1/2/3 have the chance to reset the exception  before 4, since WriterCallback#onCompletion is in asynchronous threads and assigning exception to asyncProducerException is not in the mailboxExecutor, which will expose the exception to 1/2/3 before mailboxExecutor.execute.

Reply:
As you mentioned in another comment: " If we bubble up the exception the task dies without the call-site being able to properly react to it."

I'm a little confused: 
If we hope to bubble up the exception only in main thread with mailboxExecutor.execute, the problem of [FLINK-31305](https://issues.apache.org/jira/browse/FLINK-31305) will occur again: 
in-flight checkpoint will be completed, once the mailboxExecutor.execute throw the exception a little late, then the job restarts and recovers from this error checkpoint

Throwing exception depending on mailboxExecutor.execute is not safe enough since mailboxExecutor thread and checkpoint thread are different. So I agree with the fix of FLINK-31305 bubbling up the exception in 1.writer() 2. flush() 3. close(), which can  trigger global fail over ASAP. But in FLINK-31305, the exception may be reset by mailboxExecutor.execute, 1/2/3 can not see the exception and the in-flight checkpoint with kafka sending problems may be completed. That's  why I remove the logic of resetting to null.

If there is any mistake in my understanding, please point it out to me, Thanks.  @AHeise

Reply:
Thanks for remind, I need some time to think over about this part

Reply:
@AHeise  I've remove the modification before about the reset logic in checkAsyncException(), and i understand what you mean about "we never have a chance to concurrently reset the exception" and agree with you.

Btw, I'm still little curious: if we don't reset the asyncProducerException to null in checkAsyncException(), what problems we will meet? As the comment from [FLINK-31305] : Propagate the first exception since amount of exceptions could be large.  Is this the only reason that we throw the Exception for only one time?  This seems not so persuasive for me.

Reply:
If we keep the asyncProducerException until KafkaWriter task is restarted and assign to asyncProducerException null again, that seems safer if someone modifies the related logic later...

Reply:
It's a good point to raise if we need to reset at all or not. I'm assuming @mas-chen did it to emulate the previous behavior.

From what I can see, the only real downside is that we rethrow the exception in close (which is always called also in case of errors). That could prevent some proper cleanup and will count exceptions twice in most cases, which will inflate the metrics. I think the latter point has quite a bit of side-effects:

I'm expecting most customers to have some alerts on the exceptions metrics, which could mean that it's a behavior change that would alert more users than in previous versions. So in the worst case, some poor data engineer will be paged in the middle of the night because of that. And I would expect a few alerts to be changed to be less noisy.

All in all, I think that resetting the exception would be good IF we are all sure that there won't be any exception missed. Being sure that we can't have corrupted state is of course more important than having accurate metrics but ideally we have both.

Reply:
Very good consideration for "data engineer will be paged in the middle of the night" once the alerts on the exceptions metrics changes. Then I think we need to be very cautious for the following modification about the exception logic. And the new test KafkaWriterFaultToleranceITCase can play a role for this concerns.